### ISSUE TITLE
in addition to the diff, generate a JSON file

### ISSUE THREAD

**Issue by monperrus** [SpoonLabs/sorald#193]
in addition to the diff, generate a JSON file containing the metadata about the fix, that could be used later for the explanation in the PR message.

**Comment by slarse** [comment]
As per our discussion today, we want at least the following:

* For each repair
    - The rule that was repaired
    - Location of the repair (relative from repo root filepath + line nr?)
* Failures (when a processor crashes, see #210 )
* Various statistics
    - Time to parse
    - Time to repair

Ping @khaes-kth @fermadeiral @nharrand just so you know this issue exists!

**Comment by fermadeiral** [comment]
I think it is also important to record the number of warnings per rule before and after the transformations.

**Comment by fermadeiral** [comment]
Maybe we should also record the args given as input so we can know how to reproduce something given a json file.

**Comment by slarse** [comment]
> I think it is also important to record the number of warnings per rule before and after the transformations.

That would incur heavy overhead in Sorald. Perhaps we should have an extra argument, `--gatherStats`, or something like that?

> Maybe we should also record the args given as input so we can know how to reproduce something given a json file.

That is an excellent idea! For reproducibility, we should probably include the version number as well.

**Comment by slarse** [comment]
As #220 is now merged, I'll start working on this. There may be some overlap in JSON stuff with @khaes-kth's work on #225, but we can fix that after the fact.

**Comment by slarse** [comment]
To make this modular, I'm thinking that we'll want to pass some sort of event handler to the repair process. And then, for example, when Sorald starts parsing the input, it registers an event `PARSE_START` with the event handler, and the handler can do whatever it wants with that. When parsing ends, it registers `PARSE_END`, and that way we can time the parse time.

Similarly, we could pass `VIOLATION_DETECTED` when a violation is detected, and `VIOLATION_REPAIRED` (along with more info) whenever a violation is repaired. We can also squeeze in stuff like `PROCESSOR_CRASH` to cover #210. 

I haven't quite thought this through all the way. For example, different events will require different metainfo (perhaps a simple hashmap will do for that). But I think this could be a decent idea. I plan to work on this tomorrow, so hopefully I'll have a PR with a suggestion by the end of the day.

**Comment by slarse** [comment]
To give an extremely rough idea of what I'm thinking, create an event handler like so and pass it to the `repair` method:

```java
            SoraldEventHandler eventHandler =
                    new SoraldEventHandler() {
                        private long parseStart = -1;
                        private long parseEnd = -1;

                        @Override
                        public void event(EventType type) {
                            switch (type) {
                                case PARSE_START:
                                    parseStart = System.nanoTime();
                                    break;
                                case PARSE_END:
                                    parseEnd = System.nanoTime();
                                    break;
                            }
                        }

                        @Override
                        public void close() {
                            System.out.println(
                                    "Time to parse: "
                                            + (parseEnd - parseStart) / 1_000_000
                                            + " ms");
                        }
                    };
            Stream<CtModel> models = repair(inputDir, processor, eventHandler)
            // process the models stream
            eventHandler.close();
```

And then register events inside of the repair process like so:

```java
    CtModel defaultRepair(
            Path inputDir, SoraldAbstractProcessor<?> processor, SoraldEventHandler eventHandler) {
        eventHandler.event(EventType.PARSE_START);

        Launcher launcher = new Launcher();
        launcher.addInputResource(inputDir.toString());
        CtModel model = initLauncher(launcher).getModel();

        eventHandler.event(EventType.PARSE_END);

        File inputBaseDir = FileUtils.getClosestDirectory(inputDir.toFile());
        processor.initResource(inputDir.toString(), inputBaseDir);
        repairModelWithInitializedProcessor(model, processor);

        return model;
    }
```

Of course, we wouldn't print the results to stdout, but it's just the basic idea.

**Comment by monperrus** [comment]
closed per #232
